### PR TITLE
Stabilize Workspace Inheritance

### LINKED ISSUE #8415: Tracking issue for Deduplicate Cargo workspace information RFC
This is a tracking issue for the  RFC: Deduplicate Cargo workspace information (https://github.com/rust-lang/rfcs/pull/2906).

Documentation: https://doc.rust-lang.org/nightly/cargo/reference/unstable.html#workspace-inheritance

**Steps:**

- [ ] Implement the RFC (cc @rust-lang/cargo -- can anyone write up mentoring
      instructions?)
  - [x] Support for inheriting in root package (#10497)
  - [x] Support inheriting in any workspace member (#10517)
  - [x] Correctly inherit fields that are relative paths (#10538, #10548)
  - [x] Add  `rust-version` (not in RFC), see [epage's comment](https://github.com/rust-lang/cargo/issues/8415#issuecomment-1093362940) (https://github.com/rust-lang/cargo/pull/10563)
  - [x] Switch the inheritance source from `workspace` to `workspace.package`, see [epage's comment](https://github.com/rust-lang/cargo/issues/8415#issuecomment-1097422198) (https://github.com/rust-lang/cargo/pull/10564)
  - [x] Add `include` and `exclude` now that `workspace.package` is done, see [epage's comment](https://github.com/rust-lang/cargo/issues/8415#issuecomment-1097422198) (https://github.com/rust-lang/cargo/pull/10565)
  - [x] cargo-add support (https://github.com/rust-lang/cargo/pull/10606)
    - Non-blocking: https://github.com/rust-lang/cargo/issues/10608
  - [x] Optimizations, as needed (https://github.com/rust-lang/cargo/pull/10568)
- [x] Adjust documentation (https://github.com/rust-lang/cargo/pull/10611)
  - [x] Decide on documented style (`key = {worskspace = true}` vs `key.workspace = true`)
- [ ] Stabilization PR ([see instructions on cargo-dev-guide](https://doc.crates.io/contrib/process/unstable.html#stabilization))

**Changes from RFC**
- Not locking workspace dependencies, see [alexchrichton's comment](https://github.com/rust-lang/rfcs/pull/2906#issuecomment-1095098143)
- Not inserting a version into path dependencies when publishing, see [epage's comment](https://github.com/rust-lang/cargo/issues/8415#issuecomment-1097412783)
- Changed the table `package` fields inherit from, from `workspace` to `workspace.package`, see [epage's comment](https://github.com/rust-lang/cargo/issues/8415#issuecomment-1097422198)
- Fields
  - Explicitly not supporting `resolver`, see [epage's comment](https://github.com/rust-lang/cargo/issues/8415#issuecomment-1098262362)
  - Supporting `edition` which the RFC showed inconsistent status for
  - Supporting `rust-version` which is newer than the RFC
  - Supporting `include` and `exclude` which the RFC couldn't, see [epage's comment](https://github.com/rust-lang/cargo/issues/8415#issuecomment-1098262362)

**Unresolved questions:**
- Do we stablize inheriting from `workspace.package` and `workspace.dependencies` or do we switch which tables we use?  see [epage's comment](https://github.com/rust-lang/cargo/issues/8415#issuecomment-1097412783)
- Preferred documentation style: RFC uses `field = { workspace = true }` but TOML allows `field.workspace = true`
- Is the performance good enough?

### PR DESCRIPTION
**Pull request by Muscraft**
What does this PR try to resolve?

Stabilize Workspace Inheritance as [#8415 (comment)](https://github.com/rust-lang/cargo/issues/8415#issuecomment-1182638594) concluded.

Close #8415

### CODE REVIEW

File: src/cargo/core/features.rs
Comment (epage) [inline src/cargo/core/features.rs:417]: I'm assuming 1.64 is correct

**Review by epage — APPROVED** [review]
Looks good to me; we just need to wait for the final comment period to complete before merging

File: src/doc/src/reference/workspaces.md
Comment (ehuss) [inline src/doc/src/reference/workspaces.md:146]: I prefer to tend towards keeping examples closer to reality without too many elisions. This can help avoid confusion with new users who may misinterpret it. I think in this case, it wouldn't hurt to include a few words instead of an ellipsis. Perhaps something like:

```suggestion
description = "A short description of my package"
```

File: src/doc/src/reference/workspaces.md
Comment (ehuss) [inline src/doc/src/reference/workspaces.md:147]: It doesn't look like https://github.com/example appears to be owned by someone not affiliated with GitHub.  I think it would be better to include something that is more of a standard site like example.com:

```suggestion
documentation = "https://example.com/bar"
```

File: src/doc/src/reference/workspaces.md
Comment (ehuss) [inline src/doc/src/reference/workspaces.md:180]: Similar to above, I think it is good to use real-world examples (and particularly ones that are either managed by rust-lang or other well-trusted maintainers).  I realize this is trying to clearly illustrate that it can be shared across different dependency kinds, but I think the example below makes that pretty clear.  I'd also like to avoid malicious squatters on these package names. Perhaps something like this?

```suggestion
cc = "1.0.73"
rand = "0.8.5"
regex = { version = "1.6.0", default-features = false, features = ["std"] }
```

File: src/doc/src/reference/workspaces.md
Comment (ehuss) [inline src/doc/src/reference/workspaces.md:196]:
```suggestion
[dependencies]
regex = { workspace = true, features = ["unicode"] }

[build-dependencies]
cc.workspace = true

[dev-dependencies]
rand.workspace = true
```

File: src/doc/src/reference/specifying-dependencies.md
Comment (ehuss) [inline src/doc/src/reference/specifying-dependencies.md:461]: Here `dep` may not be clear that this is a placeholder for what the user actually enters. Perhaps just not mention it? I think the example below is what most users will actually look at, which should make it clear what the syntax looks like. Maybe like this:

```suggestion
After that add it to the `[dependencies]` table with `workspace = true`.
```

File: src/doc/src/reference/specifying-dependencies.md
Comment (ehuss) [inline src/doc/src/reference/specifying-dependencies.md:463]: This wasn't entirely clear to me what it was meaning. Perhaps the following might be clearer?

```suggestion
Along with the `workspace` key, dependencies can also include these keys:
```

File: src/doc/src/reference/specifying-dependencies.md
Comment (ehuss) [inline src/doc/src/reference/specifying-dependencies.md:475]: I'm a little concerned with trying to keep this list up-to-date over time, and in general isn't necessary for the user to know exactly which keys aren't allowed. Perhaps this could be simplified with the following?

```suggestion
Dependencies with the `workspace=true` key cannot use any other dependency key (such as `version` or `default-features`).
```

File: src/doc/src/reference/specifying-dependencies.md
Comment (ehuss) [inline src/doc/src/reference/specifying-dependencies.md:485]: Similar to the other example, can this use realistic examples?

**Review by ehuss — COMMENTED** [review]
Thanks, this looks great!  Sorry I'm a bit nit-picky with docs, though! 😄

File: src/doc/src/reference/specifying-dependencies.md
Comment (Muscraft) [reply src/doc/src/reference/specifying-dependencies.md]: I used what you put here. I was trying to say "After that add it to the `[dependencies]` table with `{dependency name}.workspace = true`."

File: src/doc/src/reference/specifying-dependencies.md
Comment (Muscraft) [reply src/doc/src/reference/specifying-dependencies.md]: Do you think I should remove the table in `workspaces.md` as well?

File: src/doc/src/reference/specifying-dependencies.md
Comment (Muscraft) [reply src/doc/src/reference/specifying-dependencies.md]: I used the example you gave for `workspaces.md` here as well since it fit

Comment (ehuss) [reply src/doc/src/reference/specifying-dependencies.md]: I think it's fine to keep that table, as it is important for the user to know which keys they can use. We'll just need to remember to update it when any new inherited keys are added.

File: src/doc/src/reference/specifying-dependencies.md
Comment (epage) [reply src/doc/src/reference/specifying-dependencies.md]: I assume this was meant to be here
https://github.com/rust-lang/cargo/pull/10859#discussion_r924002389
> I think it's fine to keep that table, as it is important for the user to know which keys they can use. We'll just need to remember to update it when any new inherited keys are added.

**Comment by epage** [comment]
@bors r+

FCP is done and all comments seem to be resolved.  We can always iterate further on docs

Thank you @Muscraft for all of your work on this through the entire process!

**Pull request merged**
